diff --git a/esp32/lib/rlottie/src/lottie/lottieitem.cpp b/esp32/lib/rlottie/src/lottie/lottieitem.cpp
index c70f327..3f123f3 100644
--- a/esp32/lib/rlottie/src/lottie/lottieitem.cpp
+++ b/esp32/lib/rlottie/src/lottie/lottieitem.cpp
@@ -365,7 +365,8 @@ bool renderer::Layer::resolveKeyPath(LOTKeyPath &keyPath, uint32_t depth,
     if (!keyPath.skip(name())) {
         if (keyPath.fullyResolvesTo(name(), depth) &&
             transformProp(value.property())) {
-            //@TODO handle propery update.
+            if (!mFilterData) mFilterData = std::make_unique<model::FilterData>();
+            mFilterData->addValue(value);
         }
     }
     return true;
diff --git a/esp32/lib/rlottie/src/lottie/lottieitem.h b/esp32/lib/rlottie/src/lottie/lottieitem.h
index 1a50ec8..a8b814e 100644
--- a/esp32/lib/rlottie/src/lottie/lottieitem.h
+++ b/esp32/lib/rlottie/src/lottie/lottieitem.h
@@ -250,7 +250,13 @@ protected:
     inline int     frameNo() const { return mFrameNo; }
     inline float   combinedAlpha() const { return mCombinedAlpha; }
     inline bool    isStatic() const { return mLayerData->isStatic(); }
-    float opacity(int frameNo) const { return mLayerData->opacity(frameNo); }
+    float opacity(int frameNo) const
+    {
+        if (mFilterData && mFilterData->hasFilter(rlottie::Property::TrOpacity)) {
+            return mFilterData->opacity(rlottie::Property::TrOpacity, frameNo);
+        }
+        return mLayerData->opacity(frameNo);
+    }
     inline DirtyFlag flag() const { return mDirtyFlag; }
     bool             skipRendering() const
     {
@@ -261,6 +267,7 @@ protected:
     std::unique_ptr<LayerMask> mLayerMask;
     model::Layer *             mLayerData{nullptr};
     Layer *                    mParentLayer{nullptr};
+    std::unique_ptr<model::FilterData> mFilterData{nullptr};
     VMatrix                    mCombinedMatrix;
     float                      mCombinedAlpha{0.0};
     int                        mFrameNo{-1};
